include/llvm/IR/DomTreeUpdater.h

```diff
@@ -0,0 +1,181 @@
+  ///
+  /// \returns true when the Post dominator tree is up to date.
+  /// \returns false when the class does not hold a Post dominator
+  /// tree or the Post dominator tree is not up to date.
+  bool hasPendingPostDomTreeUpdates() const;
+  // Modify
+  /// Apply updates according to the UpdateStrategy.
+  ///
+  /// \param Updates The updates to perform on data structures held.
+  void applyUpdates(ArrayRef<DominatorTree::UpdateType> Updates);
+  /// Insert an Edge based on UpdateStrategy.
+  ///
+  /// \param From The BasicBlock edge starts.
+  /// \param To The BasicBlock edge ends.
+  void insertEdge(BasicBlock *From, BasicBlock *To);
+  /// Delete an Edge based on UpdateStrategy.
+  ///
+  /// \param From The BasicBlock edge starts.
+  /// \param To The BasicBlock edge ends.
+  void deleteEdge(BasicBlock *From, BasicBlock *To);
+  /// Delete a basic block based on UpdateStrategy.
+  ///
+  /// \param DelBB The basic block to delete.
+  void deleteBB(BasicBlock *DelBB);
+  /// Delete a basic block based on UpdateStrategy with callback.
+  /// The callback will be called after DelBB -> removeFromParent()
+  /// and before delete DelBB; It will deduplicate with other DelBBs.
+  ///
+  /// \param DelBB The basic block to delete.
+  /// \param Callback The callback function to be called before delete DelBB;
+  void callbackDeleteBB(BasicBlock *DelBB, function_ref<void()> Callback);
+  /// Recalculate the dominator tree.
+  /// In this scenario, it is supposed the post dominator
+  /// tree should not be recalculated.
+  /// This function should be called when the class holds a DomTree.
+  void recalculateDomTree(Function &F);
+  /// Recalculate the post dominator tree.
+  /// In this scenario, it is supposed the dominator
+  /// tree should not be recalculated.
+  /// This function should be called when the class holds a PostDomTree.
+  void recalculatePostDomTree(Function &F);
+  /// Recalculate Both of the trees.
+  /// This function should be called when the class holds both of the trees.
+  void recalculateAll(Function &F);
+  // flush
+  /// FLush DomTree updates to get DomTree up to date.
+  /// It will also clear out of date updates and unnecessary
+  /// stored BasicBlocks.
+  /// It should only be called when there is a DomTree.
+  DominatorTree &flushDomTree();
+  /// FLush PostDomTree updates to get PostDomTree up to date.
+  /// It will also clear out of date updates and unnecessary
+  /// stored BasicBlocks.
+  /// It should only be called when there is a PostDomTree.
+  PostDominatorTree &flushPostDomTree();
+  /// Flush all pending updates to get DomTree and PostDomTree up to date.
+  /// PendingUpdates and PendingDeletedBBs will be cleared.
+  /// This function can be called not knowing which tree is held.
+  void flushAll();
+  // Debug
+  /// Debug method to help view the state of the internal data structures.
+  LLVM_DUMP_METHOD void dump() const;
+
+private:
+  class CallBackOnDeletion final : public CallbackVH {
```

Comment:
Are there some other places in LLVM that subclass ValueHandles?

Reply:
Yes, there are lots of derived class of CallbackVH declared as final in the existing codebase. (e.g., BasicBlockCallbackVH declared in include/llvm/Analysis/BranchProbabilityInfo.h). I have looked into unittests/IR/ValueHandleTest.cpp when I implemented this class, and I believe it's the correct way to implement it.